### PR TITLE
feat: add estimatedGas, estimatedGasTokenRefund and minimumProtocolFee to /quote and /price  response

### PR DESCRIPTION
**Pull request by fragosti**
# Description

- Add `estimatedGas` field which represents a more optimistic gas estimate for the UI.
- Add `estimagedGasTokenRefund` field which represents an estimate for the gas token refund in the transaction.
- Add `minimumProtocolFee` field which represents the protocol fee if only the minimum amount of orders are filled.
- Did some refactoring as the main method was getting long/messy.

# Testing Instructions

It is deployed on staging.  https://staging.api.0x.org/swap/v0/quote?sellToken=DAI&buyToken=WETH&buyAmount=1000000000000000000000

# Checklist

-   [x] Update [documentation](https://github.com/0xProject/website/blob/development/mdx/api/index.mdx) as needed. **Website Documentation PR: https://github.com/0xProject/website/pull/302/files**
-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.

### CODE REVIEW

**Comment by fragosti** [comment]
deploy staging

**Comment by fragosti** [comment]
deploy-staging

**Comment by fragosti** [comment]
deploy staging

File: src/config.ts [hunk 1]
```diff
@@ -205,6 +205,10 @@ export const PROMETHEUS_PORT: number = _.isEmpty(process.env.PROMETHEUS_PORT)
     ? 8080
     : assertEnvVarType('PROMETHEUS_PORT', process.env.PROMETHEUS_PORT, EnvVarType.Port);
 
+export const GST2_WALLET_ADDRESS: string = _.isEmpty(process.env.GST2_WALLET_ADDRESS)
+    ? '0x000000d3b08566be75a6db803c03c85c0c1c5b96'
```
Comment (fragosti) [inline src/config.ts:209]: TODO: Remove this value.

File: src/utils/service_utils.ts [hunk 2]
```diff
@@ -132,4 +134,30 @@ export const serviceUtils = {
             breakdown,
         );
     },
+    getEstimatedGasTokenRefundInfo(
+        orders: OptimizedMarketOrder[],
+        gasTokenBalance: BigNumber = new BigNumber('100000000000000000000000'),
```
Comment (fragosti) [inline src/utils/service_utils.ts:139]: TODO: remove.

**Comment by fragosti** [comment]
deploy staging

File: src/constants.ts [hunk 3]
```diff
@@ -15,6 +15,7 @@ export const DEFAULT_LOGGER_INCLUDE_TIMESTAMP = true;
 export const ONE_SECOND_MS = 1000;
 export const ONE_MINUTE_MS = ONE_SECOND_MS * 60;
 export const TEN_MINUTES_MS = ONE_MINUTE_MS * 10;
+export const GST2_TOKEN_ADDRESS = '0x0000000000b3f879cb30fe243b4dfee438691c04';
```
Comment (BMillman19) [inline src/constants.ts:18]: Might this belong in the @0x/contract-addresses package? What happens on kovan / ganache?

File: src/config.ts [hunk 4]
```diff
@@ -205,6 +205,10 @@ export const PROMETHEUS_PORT: number = _.isEmpty(process.env.PROMETHEUS_PORT)
     ? 8080
     : assertEnvVarType('PROMETHEUS_PORT', process.env.PROMETHEUS_PORT, EnvVarType.Port);
 
+export const GST2_WALLET_ADDRESS: string = _.isEmpty(process.env.GST2_WALLET_ADDRESS)
```
Comment (BMillman19) [inline src/config.ts:208]: Same thing here about multiple networks

File: src/utils/service_utils.ts [hunk 5]
```diff
@@ -132,4 +134,34 @@ export const serviceUtils = {
...
+            };
+        }
+        // tslint:disable:custom-no-magic-numbers
+        const costOfBridgeFills = BigNumber.sum(...bridgeFills.map(o => GAS_SCHEDULE[o.source]))
+            .plus(bridgeFills.length * 5000)
+            .plus(20000);
+        const usedGasTokens = BigNumber.min(
+            gasTokenBalance,
+            costOfBridgeFills
+                .plus(14154)
```
Comment (BMillman19) [inline src/utils/service_utils.ts:155]: Should these numbers be in constants somewhere?

File: src/config.ts (hunk 4 above)
Comment (fragosti) [reply src/config.ts:208]: Are you saying we make it required?

File: src/constants.ts (hunk 3 above)
Comment (fragosti) [reply src/constants.ts]: The contract-addresses package mostly contains smart contracts that we own, this is just a token, so put it in the token_metadatas_for_networks file.

File: src/config.ts (hunk 4 above)
Comment (BMillman19) [reply src/config.ts:208]: just wanted to call it out, what happens if we forget to override the deployment? Any risk?

Comment (fragosti) [reply src/config.ts]: Nothing would happen but I moved it to a chainId -> string object in constants.ts which probably makes more sense

**Pull request merged**